### PR TITLE
feat: Add sanitycheck VM has HA disks

### PR DESCRIPTION
**Pull request by lperdereau**

### CODE REVIEW

File: src/pvecontrol/sanitycheck.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+class CheckCode(Enum):
+  CRIT = 'CRITICAL'
+  WARN = 'WARNING'
+  INFO = 'INFO'
+  OK = 'OK'
+
+ICONS = {
+  CheckCode.CRIT.value: '❌',
+  CheckCode.WARN.value: '⚠️',
+  CheckCode.INFO.value: 'ℹ️',
```
Comment (paullaffitte) [inline src/pvecontrol/sanitycheck.py:22]: `CheckCode.INFO` doesn't seem to be used, should we keep it anyway for a later use?

File: src/pvecontrol/sanitycheck.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+      return (allocated_mem +  min_mem) >= max_mem
+
+    # Check nodes capacity
+    def _validate_nodes_capacity(self):
+      node_config = self._proxmox.config['node']
+      check = Check(CheckType.Node, "Check Node capacity")
+      for node in self._proxmox.nodes:
+        if self._check_cpu_overcommit(node.maxcpu, node_config['cpufactor'], node.allocatedcpu):
+          msg = "Node %s is in cpu overcommit status: %s allocated but %s available"%(node.node, node.allocatedcpu, node.maxcpu)
+          check.add_messages(CheckMessage(CheckCode.CRIT, msg))
```
Comment (paullaffitte) [inline src/pvecontrol/sanitycheck.py:115]: Shouldn't this be a warning instead?

File: src/pvecontrol/sanitycheck.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+      return check
+
+    # Check HA groups
+    def _validate_ha_groups(self):
+      check = Check(CheckType.HA, "Check HA groups")
+      for group in self._ha['groups']:
+        num_nodes = len(group['nodes'].split(","))
+        if num_nodes < 2:
+          msg = f"Group {group['group']} contain only {num_nodes} node"
+          check.add_messages(CheckMessage(CheckCode.WARN, msg))
```
Comment (paullaffitte) [inline src/pvecontrol/sanitycheck.py:136]: And this be a critical? :thinking:

File: src/pvecontrol/sanitycheck.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+    def _check_mem_overcommit(self, max_mem, min_mem, allocated_mem):
+      return (allocated_mem +  min_mem) >= max_mem
+
+    # Check nodes capacity
+    def _validate_nodes_capacity(self):
+      node_config = self._proxmox.config['node']
+      check = Check(CheckType.Node, "Check Node capacity")
+      for node in self._proxmox.nodes:
+        if self._check_cpu_overcommit(node.maxcpu, node_config['cpufactor'], node.allocatedcpu):
```
Comment (paullaffitte) [inline src/pvecontrol/sanitycheck.py:113]: Not sure if it's normal, but it seems that `cpufactor` is not always present in `node_config`:

```
[pyenv:.env][17:27:24.91][pvecontrol][feat-add-sanitychecks-ha-vms *=][1]:~/dev/internal/pvecontrol$ pvecontrol -c preprod sanitycheck
INFO:root:Proxmox cluster: preprod
Traceback (most recent call last):
  File "/home/paul/dev/internal/pvecontrol/.env/bin/pvecontrol", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/paul/dev/internal/pvecontrol/src/pvecontrol/__init__.py", line 129, in main
    args.func(proxmoxcluster, args)
  File "/home/paul/dev/internal/pvecontrol/src/pvecontrol/actions/cluster.py", line 57, in action_sanitycheck
    sc.run()
  File "/home/paul/dev/internal/pvecontrol/src/pvecontrol/sanitycheck.py", line 82, in run
    self._checks.append(getattr(self, method)())
                        ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/paul/dev/internal/pvecontrol/src/pvecontrol/sanitycheck.py", line 113, in _validate_nodes_capacity
    if self._check_cpu_overcommit(node.maxcpu, node_config['cpufactor'], node.allocatedcpu):
                                               ~~~~~~~~~~~^^^^^^^^^^^^^
KeyError: 'cpufactor'
```

(If I print `node_config` I get an empty dict)

Comment (lperdereau) [reply src/pvecontrol/sanitycheck.py]: Need to be rebase i will did it.

File: src/pvecontrol/sanitycheck.py (hunk 2 above)
Comment (lperdereau) [reply src/pvecontrol/sanitycheck.py]: https://github.com/enix/pvecontrol/pull/18#discussion_r1865913828

File: src/pvecontrol/sanitycheck.py (hunk 1 above)
Comment (lperdereau) [reply src/pvecontrol/sanitycheck.py]: I will remove it

File: src/pvecontrol/utils.py [hunk 5]
```diff
@@ -2,12 +2,39 @@
...
+   UNDERLINE = '\033[4m'
+   END = '\033[0m'
+
+
+def teminal_support_colors():
+  try:
+    subprocess.run(["tput", "colors"], check=True, capture_output=True)
+  except subprocess.CalledProcessError:
+    return False
+  return True
```
Comment (plaffitt) [inline src/pvecontrol/utils.py:27]: This part of the code seems to be unused.

**Comment by lcaflc** [comment]
There is an error on a test cluster:
```
❯ python3 pvecontrol.py -c enix-proxmox-2 sanitycheck
INFO:root:Proxmox cluster: enix-proxmox-2
Traceback (most recent call last):
  File "/home/caf/GIT/Enix/products/pvecontrol/src/pvecontrol.py", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/caf/GIT/Enix/products/pvecontrol/src/pvecontrol/__init__.py", line 145, in main
    args.func(proxmoxcluster, args)
  File "/home/caf/GIT/Enix/products/pvecontrol/src/pvecontrol/actions/cluster.py", line 59, in action_sanitycheck
    exitcode = sc.run(checks=args.check)
               ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/caf/GIT/Enix/products/pvecontrol/src/pvecontrol/sanitycheck/sanitychecks.py", line 26, in run
    check.run()
  File "/home/caf/GIT/Enix/products/pvecontrol/src/pvecontrol/sanitycheck/tests/ha_vms.py", line 22, in run
    self.add_messages(self._check_cpu_ha_consistency(ha_vms))
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/caf/GIT/Enix/products/pvecontrol/src/pvecontrol/sanitycheck/tests/ha_vms.py", line 57, in _check_cpu_ha_consistency
    if vm.config['cpu'] == 'host':
       ~~~~~~~~~^^^^^^^
KeyError: 'cpu'

```

CPU is in fact an option in the API :/

**Comment by lperdereau** [comment]
@lcaflc I correct error when we access to CPU data, the placeholder is 'Default'.
I see that 'Default' is kvm64 but recently it was update to a new CPU :
> In Proxmox VE 8.2.4 the default is "x86-64-v2-AES".

**Review by lcaflc — APPROVED** [review]

**Pull request merged**